plugins.d/python.d.plugin

```diff
@@ -0,0 +1,328 @@
+                  mod.__name__ +
+                  reason[3:] +
+                  "() function. reports failure.")
+        elif reason[:13] == "configuration":
+            debug(mod.__name__,
+                  "configuration file '" +
+                  self.configs +
+                  mod.__name__ +
+                  ".conf' not found. Using defaults.")
+        elif reason[:11] == "misbehaving":
+            debug(mod.__name__, "is misbeaving. Disabling it")
+
+    def check(self):
+        for mod in self.modules:
+            try:
+                if not mod.check():
+                    self.disable_module(mod, "failed check")
+            except AttributeError:
+                self.disable_module(mod, "no check")
+            except UnboundLocalError:
+                self.disable_module(mod, "misbehaving")
+
+    def create(self):
+        for mod in self.modules:
+            try:
+                if not mod.create():
+                    self.disable_module(mod, "failed create")
+                else:
+                    chart = mod.__name__
+                    sys.stdout.write(
+                        "CHART netdata.plugin_pythond_" +
+                        chart +
+                        " '' 'Execution time for " +
+                        chart +
+                        " plugin' 'milliseconds / run' python.d netdata.plugin_python area 145000 " +
+                        str(self.timetable[mod.__name__][1]) +
+                        '\n')
+                    sys.stdout.write("DIMENSION run_time 'run time' absolute 1 1\n\n")
+                    sys.stdout.flush()
+            except AttributeError:
+                self.disable_module(mod, "no create")
+            except UnboundLocalError:
+                self.disable_module(mod, "misbehaving")
+
+    def update_module(self, mod):
+        t1 = time.time()
+        # check if it is time to execute module update() function
+        if (t1 - self.timetable[mod.__name__][0]) < self.timetable[mod.__name__][1]:
+            return
+        try:
+            if not self.first_run:
+                t = ""
+            else:
+                t=int((t1- self.timetable[mod.__name__][0]) * 1000000)
+            if not mod.update(t):
+                self.disable_module(mod, "update failed")
+                return
+        except AttributeError:
+            self.disable_module(mod, "no update")
+            return
+        except UnboundLocalError:
+            self.disable_module(mod, "misbehaving")
+            return
+        t2 = time.time()
+        if self.first_run:
+            dt = ""
+        else:
+            dt = int((t2 - self.timetable[mod.__name__][0]) * 1000000)
+        sys.stdout.write("BEGIN netdata.plugin_pythond_"+mod.__name__+" "+str(dt)+'\n')
+        sys.stdout.write("SET run_time = " + str(int((t2 - t1) * 1000)) + '\n')
+        sys.stdout.write("END\n")
+        sys.stdout.flush()
+        self.timetable[mod.__name__][0] = t1
```

Comment:
hm... I don't really understand this, so please verify these are made properly:

I have concluded that the best way to to recursive collect values is via fixed steps. So you should use just these timings:

1. `now` (the current time in ms, as taken just **before** starting the module update)
2. `last_run` (the last time in ms, a module update was called)
3. `next_run` (the next time is ms, a module update should be called). You should increase this independently on any other timing, i.e. `while(next_run < now); next_run += module.update_every;`. This means that the module may lose a step, but it will always be collected in fixed steps.
4. The time to reported in BEGIN statements should be `now - last_run` (zero on the first iteration). 
5. Once a module finishes, you should set `last_run = now` (where `now` is the time before updating the module)

To report the latency of each module (i.e. how much time did it needed to execute), you will need another (your `t2`) to calculate `dt = t2 - now`.

Sorry, my python skills are not good, so I cannot verify this is the way you have done it.